crates/transport/src/layers/cache.rs

```diff
@@ -0,0 +1,230 @@
+            trace!(?path, "Cache file does not exist.");
+            return Ok(());
+        };
+        let file = std::fs::File::open(path).map_err(TransportErrorKind::custom)?;
+        let file = BufReader::new(file);
+        let entries: Vec<FsCacheEntry> =
+            serde_json::from_reader(file).map_err(TransportErrorKind::custom)?;
+        let mut cache = cache.write();
+        for entry in entries {
+            cache.put(entry.key, entry.value);
+        }
+
+        Ok(())
+    }
+}
+
+/// Configuration for the cache layer.
+/// For future extensibility of the configurations.
+#[derive(Debug, Clone)]
+pub struct CacheConfig {
+    /// Maximum number of items to store in the cache.
+    pub max_items: usize,
+    /// Path of the cache file.
+    pub path: PathBuf,
+}
+
+impl<S> Layer<S> for CacheLayer {
+    type Service = CachingService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        CachingService::new(inner, self.clone())
+    }
+}
+
+impl Drop for CacheLayer {
+    fn drop(&mut self) {
+        let _ = self.save_cache();
+    }
+}
+
+/// Caching service.
+#[derive(Debug, Clone)]
+pub struct CachingService<S> {
+    /// Inner transport service.
+    inner: S,
+    /// Cache layer that houses the cache and it's config
+    layer: CacheLayer,
+}
+
+impl<S> CachingService<S> {
+    /// Instantiate a new cache service.
+    pub fn new(inner: S, layer: CacheLayer) -> Self {
+        Self { inner, layer }
+    }
+
+    /// Resolves a `SerializedRequest` into a `RawValue` if it exists in the cache.
+    pub fn resolve(&self, req: &SerializedRequest) -> TransportResult<Option<Box<RawValue>>> {
+        let key = req.params_hash();
+        let value = self.layer.get(&key)?;
+
+        match value {
+            Some(value) => {
+                let raw = RawValue::from_string(value).map_err(RpcError::ser_err)?;
+                Ok(Some(raw))
+            }
+            None => Ok(None),
+        }
+    }
+
+    /// Handles a cache hit.
+    fn handle_cache_hit(&self, id: Id, raw: Box<RawValue>) -> ResponsePacket {
+        let payload = ResponsePayload::Success(raw);
+        let response = Response { id, payload };
+        ResponsePacket::Single(response)
+    }
+}
+
+impl<S> Service<RequestPacket> for CachingService<S>
+where
+    S: Service<RequestPacket, Response = ResponsePacket, Error = TransportError>
+        + Send
+        + 'static
+        + Clone,
+    S::Future: Send + 'static,
+{
+    type Response = ResponsePacket;
+    type Error = TransportError;
+    type Future = TransportFut<'static>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), TransportError>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: RequestPacket) -> Self::Future {
+        let mut inner = self.inner.clone();
+        let this = self.clone();
+        match req.clone() {
+            RequestPacket::Single(ser_req) => {
+                let params_hash = ser_req.params_hash();
```

Comment:
hmm this logic has a critical flaw, because block tags like latest will result in the same hash and therefor return an outdated response.

not sure how to mitigate this